diff --git a/Makefile.all.am b/Makefile.all.am
index 94d65f2cc..c102940cc 100755
--- a/Makefile.all.am
+++ b/Makefile.all.am
@@ -175,7 +175,7 @@ AM_CPPFLAGS_@VGCONF_PLATFORM_PRI_CAPS@ = \
 	-I$(top_builddir)/include \
 	-I$(top_srcdir)/VEX/pub \
 	-I$(top_builddir)/VEX/pub \
-	-DVGA_@VGCONF_ARCH_PRI@=1 \
+	-DVGA_@VGCONF_ARCH_PRI@=0 \
 	-DVGO_@VGCONF_OS@=1 \
 	-DVGP_@VGCONF_ARCH_PRI@_@VGCONF_OS@=1 \
 	-DVGPV_@VGCONF_ARCH_PRI@_@VGCONF_OS@_@VGCONF_PLATVARIANT@=1
